lib/evmone/eof.cpp

```diff
@@ -349,40 +349,41 @@ bool validate_rjump_destinations(bytes_view code) noexcept
 std::variant<EOFValidationError, int32_t> validate_max_stack_height(
     bytes_view code, size_t func_index, const std::vector<EOFCodeType>& code_types)
 {
-    assert(!code.empty());
-
     // Special values used for detecting errors.
     static constexpr int32_t LOC_UNVISITED = -1;  // Unvisited byte.
-    static constexpr int32_t LOC_IMMEDIATE = -2;  // Immediate byte.
+
+    struct StackHeightRange
+    {
+        int32_t min = 0;
+        int32_t max = 0;
+    };
+
+    assert(!code.empty());
 
     // Stack height in the header is limited to uint16_t,
     // but keeping larger size for ease of calculation.
-    std::vector<int32_t> stack_heights(code.size(), LOC_UNVISITED);
-    stack_heights[0] = code_types[func_index].inputs;
+    std::vector<StackHeightRange> stack_heights(code.size(), {LOC_UNVISITED, LOC_UNVISITED});
+    stack_heights[0] = {code_types[func_index].inputs, code_types[func_index].inputs};
 
-    std::stack<size_t> worklist;
-    worklist.push(0);
-
-    while (!worklist.empty())
+    for (size_t i = 0; i < code.size();)
     {
-        const auto i = worklist.top();
-        worklist.pop();
-
         const auto opcode = static_cast<Opcode>(code[i]);
 
         int stack_height_required = instr::traits[opcode].stack_height_required;
         auto stack_height_change = instr::traits[opcode].stack_height_change;
 
         auto stack_height = stack_heights[i];
```

Comment:
```suggestion
        const auto stack_height = stack_heights[i];
```

Reply:
It was update on line 436, but I decided to make a separate variable `next_stack_height` for that, maybe it's cleaner.